terraform/modules/custom_metrics/custom_metrics.tf

```diff
@@ -0,0 +1,395 @@
+  }
+}
+
+resource "kubernetes_cluster_role_binding" "custom_metrics_adapter_resource_reader" {
+  count = var.use_aws ? 1 : 0
+  metadata {
+    name = "custom-metrics-adapter:resource-reader"
+  }
+  role_ref {
+    kind      = "ClusterRole"
+    api_group = "rbac.authorization.k8s.io"
+    name      = kubernetes_cluster_role.resource_reader[0].metadata[0].name
+  }
+  subject {
+    kind      = "ServiceAccount"
+    namespace = kubernetes_namespace.custom_metrics.metadata[0].name
+    name      = module.account_mapping.kubernetes_service_account_name
+  }
+}
+
+resource "kubernetes_manifest" "external_metrics_crd" {
+  count = var.use_aws ? 1 : 0
+  manifest = {
+    apiVersion = "apiextensions.k8s.io/v1beta1"
+    kind       = "CustomResourceDefinition"
+
+    metadata = {
+      name = "externalmetrics.metrics.aws"
+    }
+
+    spec = {
+      group   = "metrics.aws"
+      version = "v1alpha1"
+      names = {
+        kind     = "ExternalMetric"
+        plural   = "externalmetrics"
+        singular = "externalmetric"
+      }
+      scope = "Namespaced"
+    }
+  }
+}
+
+resource "kubernetes_deployment" "aws_custom_metrics_adapter" {
+  count = var.use_aws ? 1 : 0
+  metadata {
+    namespace = kubernetes_namespace.custom_metrics.metadata[0].name
+    name      = "custom-metrics-adapter"
+    labels    = { app = "custom-metrics-adapter" }
+  }
+  spec {
+    replicas = 1
+    selector { match_labels = { app = "custom-metrics-adapter" } }
+    template {
+      metadata {
+        name   = "custom-metrics-adapter"
+        labels = { app = "custom-metrics-adapter" }
+      }
+      spec {
+        service_account_name = module.account_mapping.kubernetes_service_account_name
+        security_context { fs_group = 65534 }
+        container {
+          name              = "custom-metrics-adapter"
+          image             = "chankh/k8s-cloudwatch-adapter:v0.10.0"
+          image_pull_policy = "Always"
+          command           = ["/adapter", "--cert-dir=/tmp", "--secure-port=6443", "--logtostderr=true", "--v=2"]
+          port {
+            container_port = 6443
+            name           = "https"
+          }
+          port {
+            container_port = 8080
+            name           = "http"
+          }
+          volume_mount {
+            name       = "temp-vol"
+            mount_path = "/tmp"
+          }
+        }
+        volume {
+          name = "temp-vol"
+          empty_dir {}
+        }
+      }
+    }
+  }
+}
+
+resource "kubernetes_service" "aws_custom_metrics_adapter" {
+  count = var.use_aws ? 1 : 0
+  metadata {
+    namespace = kubernetes_namespace.custom_metrics.metadata[0].name
+    name      = "custom-metrics-adapter"
+    labels    = { app = "custom-metrics-adapter" }
+  }
+  spec {
+    port {
+      name        = "https"
+      port        = 443
+      target_port = 6443
+    }
+    port {
+      name        = "http"
+      port        = 80
+      target_port = 8080
+    }
```

Comment:
I believe [this page](https://kubernetes.io/docs/concepts/security/controlling-access/#api-server-ports-and-ips) gives a description of the two ports. tl;dr is: the secure port (443) is used for "normal" serving, uses TLS, has all the needed authn/authz functionality, etc; the insecure port (80) is used for turnup, bypasses a lot of security, but only binds to localhost (ie loopback) by default.

Given that last bit, I suspect that we don't actually need to expose port 80, because the metrics server won't be listening over the network on that port anyway. I tried removing it from my dev-env deployment and things seem to work OK (at least, the metric adapter logs look fine, and the horizontal-pod-autoscalers can still get their metrics). So I dropped port 80 entirely.

[and to close the loop on why it was there in the first place: I copied it blindly from [Amazon's suggested YAML config](https://raw.githubusercontent.com/amazon-archives/k8s-cloudwatch-adapter/7b73805a92a4c8dbf45b73125e8d535c867e3029/deploy/adapter.yaml)]